### PR TITLE
More general form of valid hit in pixel function

### PR DESCRIPTION
**Pull request by dildick**
Functions that check if there are valid hits in the second and third pixel (barrel/endcap) layer.

### CODE REVIEW

File: DataFormats/TrackReco/interface/HitPattern.h [hunk 1]
```diff
@@ -264,6 +264,10 @@ class HitPattern
 
     bool hasValidHitInFirstPixelBarrel() const; // has valid hit in PXB layer 1
     bool hasValidHitInFirstPixelEndcap() const; // has valid hit in PXF layer 1
+    bool hasValidHitInSecondPixelBarrel() const; // has valid hit in PXB layer 2
+    bool hasValidHitInSecondPixelEndcap() const; // has valid hit in PXF layer 2
+    bool hasValidHitInThirdPixelBarrel() const; // has valid hit in PXB layer 3
+    bool hasValidHitInThirdPixelEndcap() const; // has valid hit in PXF layer 3
```
Comment (slava77) [inline DataFormats/TrackReco/interface/HitPattern.h]: where is the implementation?

Comment (dildick) [inline DataFormats/TrackReco/interface/HitPattern.h]: Forgot to add in first commit iteration.

**Comment by slava77** [comment]
We need 81X PR.

@VinInn @rovere should comment if these new functions are practically useful.

Current implementation is a rather simplistic copy-paste, it should be replaced with less redundant code.

**Comment by makortel** [comment]
If deemed useful, what about the fourth barrel layer in the phase1 pixel? Note that the function for the third FPIX layer is already specific for >= phase1.

File: DataFormats/TrackReco/interface/HitPattern.h (hunk 1 above)
Comment (rovere) [inline DataFormats/TrackReco/interface/HitPattern.h]: why nplicate-functions and instead pass the layer as a parameter?

**Comment by slava77** [comment]
ping
there were comments made and we expect these to be addressed

**Comment by dildick** [comment]
Indeed, you are right. Should I work on a more general implementation?

**Comment by cvuosalo** [comment]
@dildick: Yes, please work on a better implementation.

File: DataFormats/TrackReco/src/HitPattern.cc [hunk 2]
```diff
@@ -323,30 +323,21 @@ uint16_t HitPattern::getHitPatternByAbsoluteIndex(int position) const
...
-        uint16_t pattern = getHitPatternByAbsoluteIndex(i);
-        if (pixelBarrelHitFilter(pattern) && (getLayer(pattern) == 1)
-                && validHitFilter(pattern)) {
-            return true;
-        }
-    }
-    return false;
-}
 
-bool HitPattern::hasValidHitInFirstPixelEndcap() const
```
Comment (VinInn) [inline DataFormats/TrackReco/src/HitPattern.cc]: No forget about
Please keep the old implementation w/o any further possible performace impact.
If you need a specific implementation make a new method.
Do not modify what was carefully crafted.

**Comment by VinInn** [comment]
After a further review ok by TRK-POG

**Comment by cvuosalo** [comment]
@cmsbuild please test

File: DataFormats/TrackReco/interface/HitPattern.h [hunk 3]
```diff
@@ -262,9 +262,9 @@ class HitPattern
     void printHitPattern(HitCategory category, int position, std::ostream &stream) const;
     void print(HitCategory category, std::ostream &stream = std::cout) const;
 
-    bool hasValidHitInFirstPixelBarrel() const; // has valid hit in PXB layer 1
-    bool hasValidHitInFirstPixelEndcap() const; // has valid hit in PXF layer 1
-
+    // has valid hit in PXB/PXF layer x 
+    bool hasValidHitInPixelLayer(bool inBarrel, uint16_t layer) const;
+
```
Comment (makortel) [inline DataFormats/TrackReco/interface/HitPattern.h]: Despite of the +1 by @VinInn let me express a concern on the interface. Using a `bool` to denote barrel/endcap is not very clear when looking the caller code alone ("Was true for barrel or for endcap?"). May I suggest to use e.g. an `enum` instead?

Comment (dildick) [inline DataFormats/TrackReco/interface/HitPattern.h]: You are right. I will use the enum in https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/DataFormats/SiPixelDetId/interface/PixelSubdetector.h.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by dildick** [comment]
Unit test error does not seem to be related to this PR: 

----- Begin Fatal Exception 29-Mar-2016 02:23:28 CEST-----------------------
An exception of category 'Key not found' occurred while
   [0] Processing run: 1 lumi: 1 event: 1
   [1] Running path 'p'
   [2] Calling event method for module TtSemiLepSignalSelMVATrainer/'buildTraintree'
   [3] Calling produce method for unscheduled module PATJetCleaner/'cleanPatJets'
   [4] Calling produce method for unscheduled module PATTauCleaner/'cleanPatTaus'
Exception Message:
pat::Tau: the ID againstElectronVLooseMVA5 can't be found in this pat::Tau.
The available IDs are: 'againstElectronLooseMVA6' 'againstElectronMVA6Raw' 'againstElectronMVA6category' 'againstElectronMediumMVA6' 'againstElectronTightMVA6' 'againstElectronVLooseMVA6' 'againstElectronVTightMVA6' 'againstMuonLoose3' 'againstMuonTight3' 'byCombinedIsolationDeltaBetaCorrRaw3Hits' 'byIsolationMVArun2v1DBdR03oldDMwLTraw' 'byIsolationMVArun2v1DBnewDMwLTraw' 'byIsolationMVArun2v1DBoldDMwLTraw' 'byIsolationMVArun2v1PWdR03oldDMwLTraw' 'byIsolationMVArun2v1PWnewDMwLTraw' 'byIsolationMVArun2v1PWoldDMwLTraw' 'byLooseCombinedIsolationDeltaBetaCorr3Hits' 'byLooseIsolationMVArun2v1DBdR03oldDMwLT' 'byLooseIsolationMVArun2v1DBnewDMwLT' 'byLooseIsolationMVArun2v1DBoldDMwLT' 'byLooseIsolationMVArun2v1PWdR03oldDMwLT' 'byLooseIsolationMVArun2v1PWnewDMwLT' 'byLooseIsolationMVArun2v1PWoldDMwLT' 'byMediumCombinedIsolationDeltaBetaCorr3Hits' 'byMediumIsolationMVArun2v1DBdR03oldDMwLT' 'byMediumIsolationMVArun2v1DBnewDMwLT' 'byMediumIsolationMVArun2v1DBoldDMwLT' 'byMediumIsolationMVArun2v1PWdR03oldDMwLT' 'byMediumIsolationMVArun2v1PWnewDMwLT' 'byMediumIsolationMVArun2v1PWoldDMwLT' 'byPhotonPtSumOutsideSignalCone' 'byTightCombinedIsolationDeltaBetaCorr3Hits' 'byTightIsolationMVArun2v1DBdR03oldDMwLT' 'byTightIsolationMVArun2v1DBnewDMwLT' 'byTightIsolationMVArun2v1DBoldDMwLT' 'byTightIsolationMVArun2v1PWdR03oldDMwLT' 'byTightIsolationMVArun2v1PWnewDMwLT' 'byTightIsolationMVArun2v1PWoldDMwLT' 'byVLooseIsolationMVArun2v1DBdR03oldDMwLT' 'byVLooseIsolationMVArun2v1DBnewDMwLT' 'byVLooseIsolationMVArun2v1DBoldDMwLT' 'byVLooseIsolationMVArun2v1PWdR03oldDMwLT' 'byVLooseIsolationMVArun2v1PWnewDMwLT' 'byVLooseIsolationMVArun2v1PWoldDMwLT' 'byVTightIsolationMVArun2v1DBdR03oldDMwLT' 'byVTightIsolationMVArun2v1DBnewDMwLT' 'byVTightIsolationMVArun2v1DBoldDMwLT' 'byVTightIsolationMVArun2v1PWdR03oldDMwLT' 'byVTightIsolationMVArun2v1PWnewDMwLT' 'byVTightIsolationMVArun2v1PWoldDMwLT' 'byVVTightIsolationMVArun2v1DBdR03oldDMwLT' 'byVVTightIsolationMVArun2v1DBnewDMwLT' 'byVVTightIsolationMVArun2v1DBoldDMwLT' 'byVVTightIsolationMVArun2v1PWdR03oldDMwLT' 'byVVTightIsolationMVArun2v1PWnewDMwLT' 'byVVTightIsolationMVArun2v1PWoldDMwLT' 'chargedIsoPtSum' 'chargedIsoPtSumdR03' 'decayModeFinding' 'decayModeFindingNewDMs' 'footprintCorrection' 'footprintCorrectiondR03' 'neutralIsoPtSum' 'neutralIsoPtSumWeight' 'neutralIsoPtSumWeightdR03' 'neutralIsoPtSumdR03' 'photonPtSumOutsideSignalCone' 'photonPtSumOutsideSignalConedR03' 'puCorrPtSum' .
----- End Fatal Exception -------------------------------------------------

**Comment by cvuosalo** [comment]
+1

For #13744 886144c8379b9a554adb33d6a09cd3d27274f81a

Generalizing the function that checks for valid pixel hits. There should be no change in monitored quantities.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_8_0_X_2016-03-27-1100 show no significant differences, as expected.

**Comment by dildick** [comment]
@cmsbuild ping

**Comment by deguio** [comment]
please test

**Comment by slava77** [comment]
@cmsbuild please test

@dildick is there 81X PR?

**Comment by dildick** [comment]
@slava77 Not yet.

**Comment by slava77** [comment]
On 4/11/16 4:36 PM, Sven Dildick wrote:
> @slava77 <https://github.com/slava77> Not yet.
> 

Hi Sven,

apart for emergencies and other rare exceptions, new features are to be
integrated
in the development release first.
81X is the development release.


Slava

**Comment by dildick** [comment]
Shall I close this one?

**Pull request closed without merging**